### PR TITLE
INTMDB-251: Update search rs and ds to use go-client v0.12.0

### LINKED ISSUE #545: Error: error getting search index information: json: cannot unmarshal array into Go struct field IndexMapping.mappings.fields of type mongodbatlas.IndexField
### Terraform CLI and Terraform MongoDB Atlas Provider Version

```hcl
Terraform v1.0.4
on darwin_amd64
+ provider registry.terraform.io/cloudamqp/cloudamqp v1.9.2
+ provider registry.terraform.io/hashicorp/aws v3.24.1
+ provider registry.terraform.io/hashicorp/google v3.81.0
+ provider registry.terraform.io/hashicorp/null v3.0.0
+ provider registry.terraform.io/hashicorp/random v3.1.0
+ provider registry.terraform.io/hashicorp/template v2.2.0
+ provider registry.terraform.io/heroku/heroku v4.1.1
+ provider registry.terraform.io/jianyuan/sentry v0.6.0
+ provider registry.terraform.io/mongodb/mongodbatlas v1.0.0
+ provider registry.terraform.io/redislabs/rediscloud v0.2.1
```

### Terraform configuration

```hcl
resource "mongodbatlas_search_index" "skus" {
  project_id      = mongodbatlas_project.main.id
  cluster_name    = mongodbatlas_cluster.main.name
  analyzer        = "lucene.standard"
  collection_name = "skus"
  database        = var.prefix
  mappings_fields = file("./templates/mongodb/search-skus.json")
  name            = "skus.search"
}

# search-skus.json
{
  "account": {
    "type": "objectId"
  },
  "productName": {
    "type": "autocomplete"
  },
  "sku": {
    "type": "autocomplete"
  },
  "variantName": {
    "type": "autocomplete"
  }
}
```

### Atlas search index definition

```hcl
{
  "mappings": {
    "dynamic": false,
    "fields": {
      "account": {
        "type": "objectId"
      },
      "productName": {
        "type": "autocomplete"
      },
      "sku": [
        {
          "analyzer": "lucene.keyword",
          "searchAnalyzer": "lucene.keyword",
          "type": "string"
        },
        {
          "type": "autocomplete"
        }
      ],
      "variantName": {
        "type": "autocomplete"
      }
    }
  }
}
```

### Steps to Reproduce
1. `terraform apply`

### Output error

```hcl
│ Error: error getting search index information: json: cannot unmarshal array into Go struct field IndexMapping.mappings.fields of type mongodbatlas.IndexField
│ 
│   with mongodbatlas_search_index.skus,
│   on mongodb.tf line 82, in resource "mongodbatlas_search_index" "skus":
│   82: resource "mongodbatlas_search_index" "skus" {
```

Please note that the index has been manually updated via Atlas UI to get that index array definition.

### PR DESCRIPTION
**Pull request by abner-dou**
## Description

Search Index Resource and Data Source was updated in order to use the new go-client release (v0.12.0)

Link to any related issue(s):

## Type of change:

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update
- [ ] Documentation fix/enhancement

## Required Checklist:

- [ ] I have signed the [MongoDB CLA](https://www.mongodb.com/legal/contributor-agreement)
- [ ] I have read the Terraform contribution guidelines
- [x] I have added tests that prove my fix is effective or that my feature works per HashiCorp requirements
- [x] I have added any necessary documentation (if appropriate)
- [ ] I have run make fmt and formatted my code

## Further comments
Tests Results:
```
-----------------------------------------------------
--- PASS: TestAccResourceMongoDBAtlasSearchIndex_withMapping (589.53s)
PASS
coverage: 8.4% of statements
ok      github.com/mongodb/terraform-provider-mongodbatlas/mongodbatlas 592.030s        coverage: 8.4% of statements
?       github.com/mongodb/terraform-provider-mongodbatlas/version      [no test files]


--- PASS: TestAccResourceMongoDBAtlasSearchIndex_basic (701.34s)
PASS
coverage: 7.7% of statements
ok      github.com/mongodb/terraform-provider-mongodbatlas/mongodbatlas 703.770s        coverage: 7.7% of statements
?       github.com/mongodb/terraform-provider-mongodbatlas/version      [no test files]

```

### CODE REVIEW

File: go.sum [hunk 1]
```diff
@@ -977,6 +977,8 @@ go.etcd.io/etcd v0.0.0-20191023171146-3cf2f69b5738/go.mod h1:dnLIgRNXwCJa5e+c6mI
 go.etcd.io/etcd v0.0.0-20200513171258-e048e166ab9c/go.mod h1:xCI7ZzBfRuGgBXyXO6yfWfDmlWd35khcWpUa4L0xI/k=
 go.mongodb.org/atlas v0.11.0 h1:fVVh+lFlqpGaUw2GhtETnQehV4ftucXksx6CRVdE+hI=
 go.mongodb.org/atlas v0.11.0/go.mod h1:MMWDsc2akjTDSG4tVQrxv/82p3QbBnqeELbtTl45sbg=
+go.mongodb.org/atlas v0.12.0 h1:/vnHX3rh8jdPrP8mRznuU/2VrGH+cCdz8/Esrzpvaus=
+go.mongodb.org/atlas v0.12.0/go.mod h1:wVCnHcm/7/IfTjEB6K8K35PLG70yGz8BdkRwX0oK9/M=
```
Comment (gssbzn) [inline go.sum:980]: [nit] please run `go mod tidy` to clean the old version from this file

File: mongodbatlas/resource_mongodbatlas_search_index.go [hunk 2]
```diff
@@ -48,9 +47,10 @@ func returnSearchIndexSchema() map[string]*schema.Schema {
 			Required: true,
 		},
 		"analyzers": {
-			Type:     schema.TypeList,
+			Type:     schema.TypeString,
 			Optional: true,
-			Elem:     customAnalyzersSchema(),
+			//Elem:     customAnalyzersSchema(),
```
Comment (gssbzn) [inline mongodbatlas/resource_mongodbatlas_search_index.go]: can we delete this if not needed?

File: mongodbatlas/resource_mongodbatlas_search_index.go [hunk 3]
```diff
@@ -48,9 +47,10 @@ func returnSearchIndexSchema() map[string]*schema.Schema {
 			Required: true,
 		},
 		"analyzers": {
-			Type:     schema.TypeList,
+			Type:     schema.TypeString,
```
Comment (gssbzn) [inline mongodbatlas/resource_mongodbatlas_search_index.go]: making sure I get this right, analyzers and field mapping for terradform are now "strings" of json that we just map to the go client `interface{}` type correct? general question nothing blcoking here

**Review by gssbzn — COMMENTED** [review]
thanks for this, there are some linting erros and also I left a little suggestion to clean the dependency file

**Review by themantissa — CHANGES_REQUESTED** [review]
Looks good - just need to address Gustavo's comments and add one thing. Can you add a short upgrade guide to note the breaking changes?  Just want to make sure we call those out (basically it's the changes you made in the docs.   Once we have that we'll move forward quickly! Thank you!

File: go.sum (hunk 1 above)
Comment (abner-dou) [reply go.sum:980]: sure!

File: mongodbatlas/resource_mongodbatlas_search_index.go (hunk 2 above)
Comment (abner-dou) [reply mongodbatlas/resource_mongodbatlas_search_index.go]: sure, I totally forgot to remove it

File: mongodbatlas/resource_mongodbatlas_search_index.go (hunk 3 above)
Comment (abner-dou) [reply mongodbatlas/resource_mongodbatlas_search_index.go]: you are right!
Now the `analyzers` field will be like the `mapping_fields` field, both are JSON now (string)

File: go.sum (hunk 1 above)
Comment (gssbzn) [reply go.sum:980]: oh I think it's still there because of the realm client... we can look at that on a different PR

**Review by gssbzn — APPROVED** [review]
LGTM

File: mongodbatlas/resource_mongodbatlas_search_index.go [hunk 4]
```diff
@@ -420,203 +277,33 @@ func marshallSearchIndexMappingFields(fields map[string]matlas.IndexField) (stri
...
 	conn := meta.(*MongoDBClient).Atlas
 	projectID := d.Get("project_id").(string)
 
 	clusterName := d.Get("cluster_name").(string)
 
-	var indexMapping *map[string]matlas.IndexField
-	indexMappingResult := unmarshalSearchIndexMappingFields(d.Get("mappings_fields").(string))
-	indexMapping = &indexMappingResult
+	var indexMapping map[string]interface{}
+	indexMapping = unmarshalSearchIndexMappingFields(d.Get("mappings_fields").(string))
```
Comment (thetonymaster) [inline mongodbatlas/resource_mongodbatlas_search_index.go]: NIT: combine the last two lines into one.

File: mongodbatlas/resource_mongodbatlas_search_index.go (hunk 3 above)
Comment (thetonymaster) [inline mongodbatlas/resource_mongodbatlas_search_index.go]: can this somehow be validated to see if there's a way to validate if it's a working JSON, can you check what AWS does in iam policies for example

**Review by thetonymaster — APPROVED** [review]

File: mongodbatlas/resource_mongodbatlas_search_index.go (hunk 3 above)
Comment (abner-dou) [reply mongodbatlas/resource_mongodbatlas_search_index.go]: I think this is achieved when the resource internally try to unmarshall the string into a map of interfaces to verify if it is different from state (this return an error when it cannot be unmarshalled) 

```
func validateSearchAnalyzersDiff(k, old, newStr string, d *schema.ResourceData) bool {
	var j, j2 interface{}

	if old == "" {
		old = "{}"
	}

	if newStr == "" {
		newStr = "{}"
	}

	if err := json.Unmarshal([]byte(old), &j); err != nil {
		log.Printf("[ERROR] cannot unmarshal old search index analyzer json %v", err)
	}
	if err := json.Unmarshal([]byte(newStr), &j2); err != nil {
		log.Printf("[ERROR] cannot unmarshal new search index analyzer json %v", err)
	}
	if diff := deep.Equal(&j, &j2); diff != nil {
		log.Printf("[DEBUG] deep equal not passed: %v", diff)
		return false
	}

	return true
}
```

File: mongodbatlas/resource_mongodbatlas_search_index.go (hunk 4 above)
Comment (abner-dou) [reply mongodbatlas/resource_mongodbatlas_search_index.go]: this was already changed

**Review by themantissa — APPROVED** [review]
Thank you so much for quick turn around!  LGTM!

**Pull request merged**